<?xml version="1.0"?>

<launch>
    <!-- 启动roscore -->
    <master auto="start"/>

    <!-- TF setting 坐标变换-->
    <include file="$(find art_racecar)/launch/includes/art_car_tf.launch.xml" />

    <!-- SENSOR DEVICE -->
    <!-- second startup ls01g lidar激光雷达 -->
    <include file="$(find art_racecar)/launch/ls01g_lidar.launch"/>
    <!-- IMU 陀螺仪-->
    <include file="$(find art_imu)/launch/imu.launch" />
    <!-- Car 底盘驱动-->
    <node pkg="art_driver" type="art_driver_node" name="art_driver_node"/>

	
      <!-- Rviz -->
    <!--node pkg="rviz" type="rviz" name="rviz" args="-d $(find art_racecar)/launch/rviz/gmapping.rviz" if="$(arg use_rviz)"/-->

        <arg name="use_rviz" default="false" />

    <!-- for amcl 初始坐标-->    
    <arg name="init_x" default="0.0" />
    <arg name="init_y" default="0.0" />
    <arg name="init_a" default="0.0" />

    <!-- Map server -->
    <node name="map_server" pkg="map_server" type="map_server" args="$(find art_racecar)/map/mymap.yaml"/>

   

    <!-- ODOMETRY -->
    <!--rf2o_Laser_Odometry 激光雷达信息估计小车运动状态  激光里程记  -->
    <include file="$(find art_racecar)/launch/includes/rf2o.launch.xml" />
    <!-- Robot_Localization 使用imu和激光雷达的数据共同完成定位-->
    <node pkg="robot_localization" type="ekf_localization_node" name="ekf_se" clear_params="true">
        <rosparam command="load" file="$(find art_racecar)/param/ekf_params.yaml" />
    </node>

    <!-- Localization -->
    <!-- AMCL 概率定位-->
    <include file="$(find art_racecar)/launch/includes/amcl.launch.xml">
        <arg name="init_x" value="$(arg init_x)"/>
        <arg name="init_y" value="$(arg init_y)"/>
        <arg name="init_a" value="$(arg init_a)"/>
    </include>

    <!-- Navstack 导航堆栈-->
    <node pkg="move_base" type="move_base" respawn="false" name="move_base">
        <!-- local planner -->
         <!--
        <param name="base_global_planner" value="global_planner/GlobalPlanner"/>  
         -->
        <param name="base_global_planner" value="navfn/NavfnROS"/>  
        <param name="base_local_planner" value="dwa_local_planner/DWAPlannerROS"/>  
        
        <!--  dwa_local_planner 动态窗口方法导航  http://wiki.ros.org/dwa_local_planner This package provides an implementation of the Dynamic Window Approach to local robot navigation on a plane.Given a global plan to follow and a costmap -->
        <rosparam file="$(find art_racecar)/param/dwa_local_planner_params.yaml" command="load"/>
        <!-- costmap layers -->    
        <rosparam file="$(find art_racecar)/param/local_costmap_params.yaml" command="load"/>
        <rosparam file="$(find art_racecar)/param/global_costmap_params.yaml" command="load"/> 
        <!-- move_base params -->
        <rosparam file="$(find art_racecar)/param/base_global_planner_params.yaml" command="load"/>
        <rosparam file="$(find art_racecar)/param/move_base_params.yaml" command="load"/>
        <remap from="/odom" to="/odometry/filtered"/>

       
    </node>

    <!-- L1 controller -->  
    <node pkg="art_racecar" type="art_car_controller" respawn="false" name="art_car_controller" output="screen">       
        <!-- L1 -->
        <param name="Vcmd" value="2.25" /> <!-- speed of car m/s -->        
        <!-- ESC -->
        <param name="baseSpeed" value="1670"/> <!-- pwm for motor constant speed, 1500: stop --> 
 

        <!-- Servo -->
        <param name="baseAngle" value="90.0"/> <!-- the middle pos of servo, for tt02: 87, for hsp: ? -->
        <param name="AngleGain" value="-4"/> <!-- for tt02: >0, for hsp: <0 -->	
        <param name="GasGain" value="2.0"/>

        <!-- remap from="/odometry/filtered" to="odom" / --> 
        <remap from="/move_base_node/NavfnROS/plan" to="/move_base/NavfnROS/plan" /> 
    </node>

    <!-- Rviz -->
    <node pkg="rviz" type="rviz" name="rviz" args="-d $(find art_racecar)/rviz/amcl.rviz" if="$(arg use_rviz)" />


</launch>